### PR TITLE
Update repo to use go:embed

### PR DESCRIPTION
**Pull request by kisunji**
### Summary
1. Replace bindata packages with stdlib go:embed
2. Modernize some `uiserver` code with newer interfaces introduced in go 1.16 (mainly working with fs.File instead of http.File) - functionality should stay the same
3. Remove steps that are no longer used from our build files
4. Add Github Action to detect diffs in `agent/uiserver/dist`

### Tip for reviewers
The `agent/uiserver/dist` files are compiled from `make ui` and can be minimized / treated as compiled assets

### CODE REVIEW

**Comment by eculver** [comment]
> Should we prevent users from committing to `agent/uiserver/dist`?

I think so. We could add `dist` to `.gitignore`. This would also mean "Add check in case `agent/uiserver/dist` has a diff but not `ui`" is no longer necessary. Since building the binary will now depend on files in `dist`, we can require `make ui` to be a prerequisite of the build targets (`consul-docker` and `ui-docker`). 

What do you think of this approach?

**Comment by johncowen** [comment]
FYI from what I remember, there was a distinct desire keep the ui artifact commited in the git repo so that folks didn't have to rebuild the UI every time they wanted to build the binary. I'm not sure if that still stands though.

I'm not entirely sure when reading the comments here what goes in `dist`, or the implications of most of this PR 😅 , so ignore me if I've missed whats going on here, or ping me offline if you want, just thought I'd mention that ^ so you knew.

**Comment by kisunji** [comment]
> 
> 
> > Should we prevent users from committing to `agent/uiserver/dist`?
> 
> I think so. We could add `dist` to `.gitignore`. This would also mean "Add check in case `agent/uiserver/dist` has a diff but not `ui`" is no longer necessary. Since building the binary will now depend on files in `dist`, we can require `make ui` to be a prerequisite of the build targets (`consul-docker` and `ui-docker`).
> 
> What do you think of this approach?

I think adding `dist` to .gitignore would prevent our CI from also committing to it. Is there a way around that?
Otherwise I thought a pre-receive hook to diff that directory and reject commits might be useful (but unsure if used across hashicorp).

I'll apply the changes for the build targets!

**Comment by kisunji** [comment]
> 
> 
> FYI from what I remember, there was a distinct desire keep the ui artifact commited in the git repo so that folks didn't have to rebuild the UI every time they wanted to build the binary. I'm not sure if that still stands though.
> 
> I'm not entirely sure when reading the comments here what goes in `dist`, or the implications of most of this PR 😅 , so ignore me if I've missed whats going on here, or ping me offline if you want, just thought I'd mention that ^ so you knew.

Yeah I had a mini discussion around that on slack as well. We will continue using CI to commit the `ui` artifacts but as the raw files under `agent/uiserver/dist` instead of being converted to bytes and dumped into a `bindata_assetfs.go` file.

**Comment by kisunji** [comment]
> 
> Since building the binary will now depend on files in `dist`, we can require `make ui` to be a prerequisite of the build targets (`consul-docker` and `ui-docker`).
> 

I thought about this and realized the `dist` files will always be present after this PR is merged. I think the added runtime cost of `ui` as a prerequisite is not worth the check (`go build` will fail at compile time anyways if `dist` doesn't exist)

File: agent/uiserver/uiserver.go
Comment (kisunji) [inline agent/uiserver/uiserver.go:104]: Effectively the same behaviour, relying on `os.Open`:
[`http.Dir`](https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/net/http/fs.go;l=71) vs [`os.DirFS`](https://cs.opensource.google/go/go/+/refs/tags/go1.16.7:src/os/file.go;l=636)

File: agent/uiserver/uiserver.go
Comment (kisunji) [inline agent/uiserver/uiserver.go:108]: we were passing around `[]bytes` and `FileInfo` to reconstruct a file later; refactored to do that work in this func and pass pointer to `fs.File` around

File: agent/uiserver/uiserver.go
Comment (kisunji) [inline agent/uiserver/uiserver.go:206]: ioutil deprecated; same call under the hood

File: agent/uiserver/buffered_file.go
Comment (kisunji) [inline agent/uiserver/buffered_file.go:8]: `fs.File` is a smaller interface than `http.File`, making it more composable in the future

File: agent/uiserver/README.MD
Comment (eculver) [inline agent/uiserver/README.MD:1]: suggest renaming the file to `README.md` (lowercase extension) for consistency.

File: .circleci/config.yml
Comment (dnephin) [inline .circleci/config.yml:682]: Now that we have multiple files and directories, not just a single file, I think we'll need to first `git rm -r agent/uiserver/dist` before we build anything to remove any stale files, right?

Otherwise we'll never be able to prune files. And I suspect since many of the files have fingerprints in the name we'd end up with a new copy of all the files every time.

Or is that remove done somewhere else that I missed?

In the removed job we have `mv pkg/packages/consul-ui/dist pkg/web_ui`. Do we need something similar to move these files over?

**Review by dnephin — COMMENTED** [review]
Thanks for working on this. Lots of nice cleanup here! I left a couple questions inline. 

Not strictly necessary, but I wonder if it would be possible to omit the committed artifact from this PR so that only the automation commits it once this is merged into main. We could also test that flow by editing the `.circleci/config.yml` and having it run on some branch to preview what that will look like.

My other main question is about local workflow. I'm not at all familiar with building the UI locally. @johncowen and @kaxcode , would you like to try out this branch locally to make sure it still behaves the way you would expect for local dev?

File: build-support/docker/Build-UI.dockerfile
Comment (kisunji) [reply build-support/docker/Build-UI.dockerfile:4]: I was having issues running `make ui` locally and the error messages hinted that I should bump the version. But now that I think about it I'm wondering how CI was building it fine 🤔

**Comment by kisunji** [comment]
> Not strictly necessary, but I wonder if it would be possible to omit the committed artifact from this PR so that only the automation commits it once this is merged into main. We could also test that flow by editing the `.circleci/config.yml` and having it run on some branch to preview what that will look like.
> 
That makes good sense! Will cut the noise out of this PR too

File: build-support/docker/Build-UI.dockerfile
Comment (david-yu) [inline build-support/docker/Build-UI.dockerfile:1]: Sorry I couldn't help but lurk, it looks like we probably want to bump to 3.14 for all Dockerfiles since 3.13 has CVEs

File: .circleci/config.yml
Comment (kisunji) [reply .circleci/config.yml:682]: Ah good catch! I thought my changes in `build-support/functions/20-build.sh` would move the files but we don't run any `make` commands that end up calling that function

File: build-support/functions/30-release.sh
Comment (kisunji) [inline build-support/functions/30-release.sh:468]: todo: missing git add step here

File: .circleci/config.yml
Comment (kisunji) [reply .circleci/config.yml:682]: Now I'm a little confused as to how those build-support/ scripts relate to our CircleCI workflows... 😕

File: build-support/docker/Build-UI.dockerfile
Comment (kisunji) [reply build-support/docker/Build-UI.dockerfile:1]: More eyes always appreciated 👍

File: build-support/docker/Build-UI.dockerfile
Comment (johncowen) [reply build-support/docker/Build-UI.dockerfile:4]: Hey @kisunji 

First thing I noticed here was related to this change, I'm not sure if things have changed upstream since you made this change, but I had to move the `npm` installation to use a different name/version also in order to get `make ui` working. I was gonna PR this separately, but now not so sure. Maybe we should bump to alpine 3.14 as a completely different PR (are we gonna need that bump on 1.10.x/1.9.x branch etc?), kinda up to you.

File: build-support/functions/20-build.sh
Comment (johncowen) [inline build-support/functions/20-build.sh:142]: Not 100% sure whats happening here, but I get a stray:

Screenshot 2021-09-27 at 11 09 37

After running `make ui` 🤔 not sure how thats happened? Oh wait I guess that REMOVEME.txt must have been commited somewhere?

**Review by johncowen — COMMENTED** [review]
Hey @kisunji 

Thanks for tackling this, exciting stuff!

I spotted a few things going through the ropes with this, so I left some inline comments here. But also, when I came to run `make ui` then `make dev-docker` and then running the resultant image, there seems to be a problem with the content that needs serving up.

When I try to curl the `ui/` endpoint I get:

```
curl: (18) transfer closed with 111 bytes remaining to read
```

So there's something amiss still.

Kinda semi-related but I have a thing I did a while ago for e2e testing basic Consul UI (in real not mock Consul) so we could make sure 'this is working'. Would that be useful here? I could resurrect it.

File: build-support/functions/20-build.sh
Comment (kisunji) [reply build-support/functions/20-build.sh:142]: Yep! The `dist/` directory needs to exist for this PR to compile but I didn't check in the artifacts so it's a placeholder just so that git doesn't ignore `dist/`

File: build-support/docker/Build-UI.dockerfile
Comment (kisunji) [reply build-support/docker/Build-UI.dockerfile:4]: @johncowen if you can make a separate PR for the version bumps (including the alphine) it would be appreciated! My PR isn't dependent on these changes at all; I just needed this version to build locally.

**Comment by kisunji** [comment]
> When I try to curl the `ui/` endpoint I get:
> 
> ```
> curl: (18) transfer closed with 111 bytes remaining to read
> ```
>
> (with all the correct HTML content above that error)
> 
> So there's something amiss still.

Thanks for the bug report! I'll try to replicate it on my end

> 
> Kinda semi-related but I have a thing I did a while ago for e2e testing basic Consul UI (in real not mock Consul) so we could make sure 'this is working'. Would that be useful here? I could resurrect it.

Yeah that would be much appreciated! There are some backend tests to make sure that certain features of our `uiserver` works (redirects, custom files, etc) but I wasn't sure how comprehensive it was

File: .circleci/config.yml
Comment (kisunji) [inline .circleci/config.yml:690]: This job was responsible for buiding bindata_assetfs.go and persisting it to the correct workspace path.

Comment (kisunji) [reply .circleci/config.yml:690]: Note that `pkg/packages/consul-ui/dist` is assumed to exist (it contains all the compiled ui files)

File: .circleci/config.yml
Comment (kisunji) [inline .circleci/config.yml:690]: This added step is now responsible for moving the compiled ui files to the correct workspace path

File: .github/workflows/build.yml
Comment (kisunji) [inline .github/workflows/build.yml:109]: Previously we were making an intermediate well-known path `pkg/web_ui` which was used by `make static-assets` which dumped the compiled files into `agent/uiserver`.

Now that we are removing `make static-assets` we can move files from the ui directory straight into `agent/uiserver`

File: .changelog/10996.txt
Comment (kisunji) [inline .changelog/10996.txt:2]:
```suggestion
ui: removed external dependencies for serving UI assets in favor of Go's native embed capabilities
```

File: .circleci/config.yml
Comment (kisunji) [inline .circleci/config.yml:1112]: filters were never relevant because the `requires` chain resolves to `frontend-cache` which only ran on `main`

File: .github/workflows/embedded-asset-checker.yml
Comment (rboyer) [inline .github/workflows/embedded-asset-checker.yml:33]: You could avoid the indent with:
```
if [[ -z "${dist_files}" ]]; then
  exit 0
fi
<rest unindented>
```

File: .github/workflows/embedded-asset-checker.yml
Comment (rboyer) [inline .github/workflows/embedded-asset-checker.yml:41]: You can also do something like:
```
if ! diff -r --brief <stuff> ; then
   # what to do when "diff" exits non-zero because of the ! earlier
fi
```
which avoids the intermediate var

Comment (kisunji) [reply .github/workflows/embedded-asset-checker.yml:41]: I think I tried a variant but Github Actions kept exiting with code 1 when evaluating `diff`. Will try your suggestion and see if it's different

**Review by rboyer — APPROVED** [review]
Some non-blocking suggestions but overall LGTM assuming you can get the new check to be green.

File: .github/workflows/embedded-asset-checker.yml
Comment (kisunji) [reply .github/workflows/embedded-asset-checker.yml:9]: I'm intentionally failing it for the moment, but a label might be useful.

**Review by rboyer — APPROVED** [review]

**Pull request merged**
